x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/pagerduty/IncidentEventTests.java

```diff
@@ -0,0 +1,161 @@
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.Matchers.equalTo;
+
+public class IncidentEventTests extends ESTestCase {
+
+    public void testPagerDutyXContent() throws IOException {
+
+        String serviceKey = randomAlphaOfLength(3);
+        boolean attachPayload = randomBoolean();
+        Payload payload = null;
+        if (attachPayload) {
+            payload = new Payload.Simple(Collections.singletonMap(randomAlphaOfLength(3), randomAlphaOfLength(3)));
+        }
+        String watchId = randomAlphaOfLength(3);
+        String description = randomAlphaOfLength(3);
+        String eventType = randomAlphaOfLength(3);
+        String incidentKey = rarely() ? null : randomAlphaOfLength(3);
+        String client = rarely() ? null : randomAlphaOfLength(3);
+        String clientUrl = rarely() ? null : randomAlphaOfLength(3);
+        String account = rarely() ? null : randomAlphaOfLength(3);
+
+        IncidentEventContext[] contexts = null;
+        List<IncidentEventContext> links = new ArrayList<>();
+        List<IncidentEventContext> images = new ArrayList<>();
+
+        if (randomBoolean()) {
+            int numContexts = randomIntBetween(0, 3);
+            contexts = new IncidentEventContext[numContexts];
+            for (int i = 0; i < numContexts; i++) {
+                if (randomBoolean()) {
+                    contexts[i] = IncidentEventContext.link("href", "text");
+                    links.add(contexts[i]);
+                } else {
+                    contexts[i] = IncidentEventContext.image("src", "href", "alt");
+                    images.add(contexts[i]);
+                }
+            }
+        }
+
+        HttpProxy proxy = rarely() ? null : HttpProxy.NO_PROXY;
+
+        IncidentEvent event = new IncidentEvent(description, eventType, incidentKey, client, clientUrl, account,
+            attachPayload, contexts, proxy);
+
+        XContentBuilder jsonBuilder = jsonBuilder();
+        jsonBuilder.startObject(); // since its a snippet
+        event.buildAPIXContent(jsonBuilder, ToXContent.EMPTY_PARAMS, serviceKey, payload, watchId);
+        jsonBuilder.endObject();
+        XContentParser parser = createParser(jsonBuilder);
+        parser.nextToken();
+
+        String actualServiceKey = null;
+        String actualWatchId = "watcher"; // hardcoded if the SOURCE is null
+        String actualDescription = null;
+        String actualEventType = null;
+        String actualIncidentKey = null;
+        String actualClient = null;
+        String actualClientUrl = null;
+        String actualSeverity = null;
+        List<IncidentEventContext> actualLinks = new ArrayList<>();
+        List<IncidentEventContext> actualImages = new ArrayList<>();
+        Payload actualPayload = null;
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
that's a lot of parsing code, can you use the `ObjectPath` class instead. Going through this while loop in a test is pretty tedious to understand what's going on.

```
ObjectPath objectPath = ObjectPath.createFromXContent(jsonBuilder.contentType().xContent(), BytesReference.bytes(jsonBuilder));
actualServiceKey = objectPath.evaluate(IncidentEvent.Fields.ROUTING_KEY.getPreferredName());
...
```

Reply:
oh this is way nicer. tyvm.